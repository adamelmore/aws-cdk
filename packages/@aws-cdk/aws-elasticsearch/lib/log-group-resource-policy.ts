import * as iam from '@aws-cdk/aws-iam';
import * as cdk from '@aws-cdk/core';
import * as cr from '@aws-cdk/custom-resources';

/**
 * Construction properties for LogGroupResourcePolicy
 */
export interface LogGroupResourcePolicyProps {
  /**
   * The log group resource policy name
   */
  readonly policyName: string;
  /**
   * The policy statements for the log group resource logs
   */
  readonly policyStatements: [iam.PolicyStatement];
}

/**
 * Creates LogGroup resource policies.
 */
export class LogGroupResourcePolicy extends cr.AwsCustomResource {
  constructor(scope: cdk.Construct, id: string, props: LogGroupResourcePolicyProps) {
    const policy = new iam.Policy(scope, props.policyName, {
      statements: props.policyStatements,
    });

    super(scope, id, {
      resourceType: 'Custom::CloudwatchLogResourcePolicy',
      onUpdate: {
        service: 'CloudWatchLogs',
        action: 'putResourcePolicy',
        parameters: {
          policyName: props.policyName,
          policyDocument: JSON.stringify(policy.document),
        },
        physicalResourceId: cr.PhysicalResourceId.of(id),
      },
      onDelete: {
        service: 'CloudWatchLogs',
        action: 'deleteResourcePolicy',
        parameters: {
          policyName: props.policyName,
        },
        ignoreErrorCodesMatching: '400',
      },
      policy: cr.AwsCustomResourcePolicy.fromStatements([
        new iam.PolicyStatement({
          actions: ['logs:PutResourcePolicy', 'logs:DeleteResourcePolicy'],
          // Resource Policies are global in Cloudwatch Logs per-region, per-account.
          resources: ['*'],
        }),
      ]),
    });
  }
}
