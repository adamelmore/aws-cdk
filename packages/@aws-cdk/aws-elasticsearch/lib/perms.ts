export const ES_READ_ACTIONS = [
  'es:ESHttpGet',
  'es:ESHttpHead',
];

export const ES_WRITE_ACTIONS = [
  'es:ESHttpDelete',
  'es:ESHttpPost',
  'es:ESHttpPut',
  'es:ESHttpPatch',
];

export const ES_READ_WRITE_ACTIONS = [
  ...ES_READ_ACTIONS,
  ...ES_WRITE_ACTIONS,
];

export const KEY_READ_ACTIONS = [
  'kms:Decrypt',
  'kms:DescribeKey',
];

export const KEY_WRITE_ACTIONS = [
  'kms:Encrypt',
  'kms:ReEncrypt*',
  'kms:GenerateDataKey*',
];

export const KEY_READ_WRITE_ACTIONS = [
  ...KEY_READ_ACTIONS,
  ...KEY_WRITE_ACTIONS,
];
