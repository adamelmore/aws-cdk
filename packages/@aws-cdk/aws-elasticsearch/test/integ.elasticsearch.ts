import { EbsDeviceVolumeType } from '@aws-cdk/aws-ec2';
import { App, Construct, Stack, StackProps } from '@aws-cdk/core';
import * as es from '../lib';

class TestStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    new es.Domain(this, 'Domain', {
      elasticsearchVersion: es.ElasticsearchVersion.ES_VERSION_7_1,
      clusterConfig: {
        masterNodes: 3,
        masterNodeInstanceType: 'm4.large.elasticsearch',
        dataNodes: 3,
        dataNodeInstanceType: 'm4.large.elasticsearch',
      },
      ebsOptions: {
        volumeSize: 10,
        volumeType: EbsDeviceVolumeType.GENERAL_PURPOSE_SSD,
      },
      logPublishingOptions: {
        slowSearchLogEnabed: true,
        appLogEnabled: true,
      },
      nodeToNodeEncryptionEnabled: true,
      encryptionAtRestOptions: {
        enabled: true,
      },
    });
  }
}

const app = new App();
new TestStack(app, 'cdk-integ-elasticsearch');
app.synth();
