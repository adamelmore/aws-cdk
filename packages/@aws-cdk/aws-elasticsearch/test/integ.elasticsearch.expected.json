{
  "Resources": {
    "DomainESLogGroupPolicy5373A2E8": {
      "Type": "Custom::CloudwatchLogResourcePolicy",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn"
          ]
        },
        "Create": {
          "service": "CloudWatchLogs",
          "action": "putResourcePolicy",
          "parameters": {
            "policyName": "ESLogPolicy",
            "policyDocument": {
              "Fn::Join": [
                "",
                [
                  "{\"Statement\":[{\"Action\":[\"logs:PutLogEvents\",\"logs:CreateLogStream\"],\"Effect\":\"Allow\",\"Principal\":{\"Service\":\"es.amazonaws.com\"},\"Resource\":[\"",
                  {
                    "Fn::GetAtt": [
                      "SlowSearchLogsE00DC2E7",
                      "Arn"
                    ]
                  },
                  "\",\"",
                  {
                    "Fn::GetAtt": [
                      "AppLogsC5DF83A6",
                      "Arn"
                    ]
                  },
                  "\"]}],\"Version\":\"2012-10-17\"}"
                ]
              ]
            }
          },
          "physicalResourceId": {
            "id": "ESLogGroupPolicy"
          }
        },
        "Update": {
          "service": "CloudWatchLogs",
          "action": "putResourcePolicy",
          "parameters": {
            "policyName": "ESLogPolicy",
            "policyDocument": {
              "Fn::Join": [
                "",
                [
                  "{\"Statement\":[{\"Action\":[\"logs:PutLogEvents\",\"logs:CreateLogStream\"],\"Effect\":\"Allow\",\"Principal\":{\"Service\":\"es.amazonaws.com\"},\"Resource\":[\"",
                  {
                    "Fn::GetAtt": [
                      "SlowSearchLogsE00DC2E7",
                      "Arn"
                    ]
                  },
                  "\",\"",
                  {
                    "Fn::GetAtt": [
                      "AppLogsC5DF83A6",
                      "Arn"
                    ]
                  },
                  "\"]}],\"Version\":\"2012-10-17\"}"
                ]
              ]
            }
          },
          "physicalResourceId": {
            "id": "ESLogGroupPolicy"
          }
        },
        "Delete": {
          "service": "CloudWatchLogs",
          "action": "deleteResourcePolicy",
          "parameters": {
            "policyName": "ESLogPolicy"
          },
          "ignoreErrorCodesMatching": "400"
        }
      },
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete"
    },
    "Domain66AC69E0": {
      "Type": "AWS::Elasticsearch::Domain",
      "Properties": {
        "CognitoOptions": {
          "Enabled": false
        },
        "EBSOptions": {
          "EBSEnabled": true,
          "VolumeSize": 10,
          "VolumeType": "gp2"
        },
        "ElasticsearchClusterConfig": {
          "DedicatedMasterEnabled": false,
          "InstanceCount": 1,
          "InstanceType": "r5.large.elasticsearch",
          "ZoneAwarenessEnabled": false
        },
        "ElasticsearchVersion": "7.1",
        "EncryptionAtRestOptions": {
          "Enabled": true
        },
        "LogPublishingOptions": {
          "ES_APPLICATION_LOGS": {
            "CloudWatchLogsLogGroupArn": {
              "Fn::GetAtt": [
                "AppLogsC5DF83A6",
                "Arn"
              ]
            },
            "Enabled": true
          },
          "SEARCH_SLOW_LOGS": {
            "CloudWatchLogsLogGroupArn": {
              "Fn::GetAtt": [
                "SlowSearchLogsE00DC2E7",
                "Arn"
              ]
            },
            "Enabled": true
          },
          "INDEX_SLOW_LOGS": {
            "Enabled": false
          }
        },
        "NodeToNodeEncryptionOptions": {
          "Enabled": true
        }
      },
      "DependsOn": [
        "DomainESLogGroupPolicy5373A2E8"
      ]
    },
    "SlowSearchLogsE00DC2E7": {
      "Type": "AWS::Logs::LogGroup",
      "Properties": {
        "RetentionInDays": 30
      },
      "UpdateReplacePolicy": "Retain",
      "DeletionPolicy": "Retain"
    },
    "AppLogsC5DF83A6": {
      "Type": "AWS::Logs::LogGroup",
      "Properties": {
        "RetentionInDays": 30
      },
      "UpdateReplacePolicy": "Retain",
      "DeletionPolicy": "Retain"
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleDefaultPolicyD28E1A5E": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutResourcePolicy",
                "logs:DeleteResourcePolicy"
              ],
              "Effect": "Allow",
              "Resource": "*"
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleDefaultPolicyD28E1A5E",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2"
          }
        ]
      }
    },
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "AssetParameters8ae75ec4aaae0510b0918d3a69fac5c978d780ae0d60bb94c65c7f5b4c498061S3Bucket67234880"
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters8ae75ec4aaae0510b0918d3a69fac5c978d780ae0d60bb94c65c7f5b4c498061S3VersionKey9802AE96"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters8ae75ec4aaae0510b0918d3a69fac5c978d780ae0d60bb94c65c7f5b4c498061S3VersionKey9802AE96"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn"
          ]
        },
        "Runtime": "nodejs12.x",
        "Timeout": 120
      },
      "DependsOn": [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleDefaultPolicyD28E1A5E",
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2"
      ]
    }
  },
  "Parameters": {
    "AssetParameters8ae75ec4aaae0510b0918d3a69fac5c978d780ae0d60bb94c65c7f5b4c498061S3Bucket67234880": {
      "Type": "String",
      "Description": "S3 bucket for asset \"8ae75ec4aaae0510b0918d3a69fac5c978d780ae0d60bb94c65c7f5b4c498061\""
    },
    "AssetParameters8ae75ec4aaae0510b0918d3a69fac5c978d780ae0d60bb94c65c7f5b4c498061S3VersionKey9802AE96": {
      "Type": "String",
      "Description": "S3 key for asset version \"8ae75ec4aaae0510b0918d3a69fac5c978d780ae0d60bb94c65c7f5b4c498061\""
    },
    "AssetParameters8ae75ec4aaae0510b0918d3a69fac5c978d780ae0d60bb94c65c7f5b4c498061ArtifactHash9212BF97": {
      "Type": "String",
      "Description": "Artifact hash for asset \"8ae75ec4aaae0510b0918d3a69fac5c978d780ae0d60bb94c65c7f5b4c498061\""
    }
  }
}